src/umpire/strategy/DynamicPool.cpp

```diff
@@ -13,382 +13,336 @@
 #include "umpire/Replay.hpp"
 
 #include <algorithm>
+#include <tuple>
+
+inline static std::size_t round_up(std::size_t num, std::size_t factor)
+{
+  return num + factor - 1 - (num - 1) % factor;
+}
 
 namespace umpire {
 namespace strategy {
 
+DynamicPool::DynamicPool(const std::string& name,
+                         int id,
+                         Allocator allocator,
+                         const std::size_t initial_alloc_bytes,
+                         const std::size_t min_alloc_bytes,
+                         CoalesceHeuristic coalesce_heuristic,
+                         const int align_bytes) noexcept :
+  AllocationStrategy(name, id),
+  m_allocator{allocator.getAllocationStrategy()},
+  m_min_alloc_bytes{min_alloc_bytes},
+  m_align_bytes{align_bytes},
+  m_coalesce_heuristic{coalesce_heuristic},
+  m_used_map{},
+  m_free_map{},
+  m_curr_bytes{0},
+  m_actual_bytes{initial_alloc_bytes},
+  m_highwatermark{initial_alloc_bytes}
+{
+  insertFree(m_allocator->allocate(initial_alloc_bytes), initial_alloc_bytes, true);
+}
 
-void DynamicPool::findUsableBlock(struct Block *&best, struct Block *&prev, std::size_t size) {
-  best = prev = nullptr;
-  for ( struct Block *iter = freeBlocks, *iterPrev = nullptr ; iter ; iter = iter->next ) {
-    if ( iter->size >= size && (!best || iter->size < best->size) ) {
-      best = iter;
-      prev = iterPrev;
-      if ( iter->size == size )
-        break;    // Exact match, won't find a better one, look no further
-    }
-    iterPrev = iter;
+DynamicPool::~DynamicPool()
+{
+  // Free any unused blocks
+  for (auto& rec : m_free_map) {
+    void* addr;
+    bool is_head;
+    std::tie(addr, is_head) = rec.second;
+    // Deallocate if this is a head
+    if (is_head) m_allocator->deallocate(addr);
```

Comment:
Just to be sure I understand, this destructor intentionally could leak if callers have not freed their memory before the destructor is called.  Is this the intention here?  If so, should we add a debug message or something to indicate that all blacks were not on the free map?

Reply:
Yes, the intention is to leak used blocks.

Reply:
I suppose this could also cause application errors, if they intentionally did not deallocate something, and the DynamicPool is deleted. Although, this should only happen at the end of the program if used in the intended way.

Reply:
Okay.  I feel like leaving some bread crumbs behind might be useful.  Should we `UMPIRE_ERROR`, `UMPIRE_LOG`, and/or `UMPIRE_ASSERT` here?
@jdahm, @davidbeckingsale?

Reply:
I added an `UMPIRE_LOG`. Is that sufficient?

Reply:
This is probably a good start.  Could we add the allocator name and the number of bytes being leaked to the log?  I also think we should use `UMPIRE_LOG(Error` instead of `UMPIRE_LOG(Debug` for this situation.

Reply:
This sounds reasonable. We decided that AllocationStrategy destructors _should not_ deallocate everything - this way, applications leaking memory would get correctly caught by tools like Valgrind.

Reply:
Resolving: I added an `UMPIRE_LOG(Error` that prints the number of leaks if there are any.